File: sdk/Node/cpay-signing.js
"use strict";
// Cito consumer SDK 2.0. Java-compatible UTF-8 form encoding and UTF-16 ordering.
const crypto = require("node:crypto");
function javaTrim(value) {
  // Java String.trim() with bounded linear scans, not a backtracking suffix regex.
  let start = 0, end = value.length;
  while (start < end && value.charCodeAt(start) <= 0x20) start++;
  while (end > start && value.charCodeAt(end - 1) <= 0x20) end--;
  return value.slice(start, end);
}
function sha256Hex(value) { return crypto.createHash("sha256").update(value, "utf8").digest("hex"); }
function scalar(value) {
  if (typeof value === "string") { encodeURIComponent(value); return value; }
  if (Number.isSafeInteger(value)) return String(value);
  throw new TypeError("Query values must be strings or safe integers; use strings for decimals");
}
function formEncode(value) {
  return encodeURIComponent(value).replace(/[!'()~]/g, c => "%" + c.charCodeAt(0).toString(16).toUpperCase()).replace(/%20/g, "+");
}
function canonicalQuery(query) {
  if (query == null) return "";
  const entries = query instanceof URLSearchParams ? [...query.entries()] : Array.isArray(query) ? query : Object.entries(query);
  const pairs = [];
  for (const [key, value] of entries) {
    if (typeof key !== "string") throw new TypeError("Query names must be strings");
    for (const item of Array.isArray(value) ? value : [value]) if (item != null) pairs.push([scalar(key), scalar(item)]);
  }
  pairs.sort((a, b) => a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : a[1] < b[1] ? -1 : a[1] > b[1] ? 1 : 0);
  return pairs.map(([key, value]) => `${formEncode(key)}=${formEncode(value)}`).join("&");
}
function canonicalString({ method, path, query, timestamp, nonce, body = "" }) {
  return [javaTrim(method).toUpperCase(), javaTrim(path), canonicalQuery(query), javaTrim(timestamp), javaTrim(nonce), sha256Hex(javaTrim(body))].join("\n");
}
function headerValue(value, name) {
  if (typeof value !== "string" || !value || value.length > 256 || /[\x00-\x20\x7f]/.test(value)) throw new Error(`${name} must be a bounded nonempty header value without whitespace`);
  return value;
}
function signRequest({ merchantNumber, privateKeyPem, method, path, query, body = "", timestamp = new Date().toISOString(), nonce = crypto.randomUUID(), idempotencyKey }) {
  headerValue(merchantNumber, "merchantNumber"); headerValue(timestamp, "timestamp"); headerValue(nonce, "nonce");
  if (idempotencyKey != null) headerValue(idempotencyKey, "idempotencyKey");
  const key = crypto.createPrivateKey(privateKeyPem);
  if (key.asymmetricKeyType !== "rsa" || key.asymmetricKeyDetails.modulusLength < 2048) throw new Error("RSA private key of at least 2048 bits required");
  const canonical = canonicalString({ method, path, query, timestamp, nonce, body });
  const signature = crypto.sign("sha256", Buffer.from(canonical, "utf8"), { key, padding: crypto.constants.RSA_PKCS1_PADDING });
  const headers = { "X-CPay-Merchant-Number": merchantNumber, "X-CPay-Signature-Version": "v2", "X-CPay-Timestamp": timestamp, "X-CPay-Nonce": nonce, "X-CPay-Signature": signature.toString("base64") };
  if (idempotencyKey != null) headers["X-CPay-Idempotency-Key"] = idempotencyKey;
  return headers;
}
module.exports = { javaTrim, formEncode, headerValue, sha256Hex, canonicalQuery, canonicalString, signRequest };

File: sdk/Python/cpay_signing.py
"""Cito consumer SDK 2.0: the existing CPay v2 RSA contract, not a new protocol."""
from __future__ import annotations

import base64
import hashlib
import re
import uuid
from datetime import datetime, timezone
from urllib.parse import quote_plus

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding, rsa


def java_trim(value: str) -> str:
    """Java String.trim(), NOT Python's wider Unicode strip(). Server compatibility."""
    return value.strip("".join(map(chr, range(33))))


def sha256_hex(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8", errors="strict")).hexdigest()


def _scalar(value) -> str:
    if isinstance(value, str):
        value.encode("utf-8", errors="strict")
        return value
    if type(value) is int:
        return str(value)
    raise TypeError("Query values must be strings or integers; use decimal strings for amounts")


def canonical_query(query=None) -> str:
    """Sort decoded names AND repeated values as Java UTF-16, then form-encode UTF-8.

    Accept a mapping with scalar/list values or a sequence of (name, value) pairs.
    Null values are omitted; use an empty string for an explicitly empty value.
    """
    if query is None:
        return ""
    entries = query.items() if hasattr(query, "items") else query
    pairs = []
    for key, value in entries:
        if not isinstance(key, str):
            raise TypeError("Query names must be strings")
        for item in value if isinstance(value, (list, tuple)) else [value]:
            if item is not None:
                pairs.append((_scalar(key), _scalar(item)))
    pairs.sort(key=lambda pair: (pair[0].encode("utf-16-be"), pair[1].encode("utf-16-be")))
    def encode(value):
        return quote_plus(value, safe="*-._", encoding="utf-8", errors="strict").replace("~", "%7E")
    return "&".join(f"{encode(key)}={encode(value)}" for key, value in pairs)


def canonical_string(method: str, path: str, query, timestamp: str, nonce: str, body: str = "") -> str:
    # The released Java verifier trims U+0000..U+0020 at field/body boundaries.
    # Compact JSON is recommended; whitespace inside JSON is always significant.
    return "\n".join((java_trim(method).upper(), java_trim(path), canonical_query(query),
                      java_trim(timestamp), java_trim(nonce), sha256_hex(java_trim(body))))


def header_value(value: str, name: str) -> str:
    if not isinstance(value, str) or not value or len(value) > 256 or re.search(r"[\x00-\x20\x7f]", value):
        raise ValueError(f"{name} must be a nonempty, bounded header value without whitespace")
    return value


def sign_request(merchant_number: str, private_key_pem: str, method: str, path: str,
                 query=None, body: str = "", timestamp: str | None = None,
                 nonce: str | None = None, idempotency_key: str | None = None) -> dict[str, str]:
    merchant_number = header_value(merchant_number, "merchant_number")
    timestamp = header_value(timestamp or datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"), "timestamp")
    nonce = header_value(nonce or str(uuid.uuid4()), "nonce")
    if idempotency_key is not None:
        header_value(idempotency_key, "idempotency_key")
    key = serialization.load_pem_private_key(private_key_pem.encode("utf-8"), password=None)
    if not isinstance(key, rsa.RSAPrivateKey) or key.key_size < 2048:
        raise ValueError("An RSA private key of at least 2048 bits is required")
    canonical = canonical_string(method, path, query, timestamp, nonce, body)
    signature = key.sign(canonical.encode("utf-8"), padding.PKCS1v15(), hashes.SHA256())
    headers = {"X-CPay-Merchant-Number": merchant_number, "X-CPay-Signature-Version": "v2",
               "X-CPay-Timestamp": timestamp, "X-CPay-Nonce": nonce,
               "X-CPay-Signature": base64.b64encode(signature).decode("ascii")}
    # Never silently invent a new business-operation key on each retry.
    if idempotency_key is not None:
        headers["X-CPay-Idempotency-Key"] = idempotency_key
    return headers
